File: modules/project/src/main/java/com/falsepattern/zigbrains/project/toolchain/ToolchainZLSConfigProvider.java
/*
 * Copyright 2023-2024 FalsePattern
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.falsepattern.zigbrains.project.toolchain;

import com.falsepattern.zigbrains.project.openapi.components.ZigProjectSettingsService;
import com.falsepattern.zigbrains.zig.environment.ZLSConfig;
import com.falsepattern.zigbrains.zig.environment.ZLSConfigProvider;
import com.intellij.notification.Notification;
import com.intellij.notification.NotificationType;
import com.intellij.notification.Notifications;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.project.ProjectUtil;
import lombok.val;

import java.nio.file.Files;
import java.nio.file.InvalidPathException;
import java.nio.file.Path;

public class ToolchainZLSConfigProvider implements ZLSConfigProvider {
    @Override
    public void getEnvironment(Project project, ZLSConfig.ZLSConfigBuilder builder) {
        val svc = ZigProjectSettingsService.getInstance(project);
        val state = svc.getState();
        var toolchain = state.getToolchain();
        if (toolchain == null) {
            toolchain = AbstractZigToolchain.suggest();
            if (toolchain == null) {
                return;
            }
            state.setToolchain(toolchain);
        }
        val projectDir = ProjectUtil.guessProjectDir(project);
        val oEnv = toolchain.zig().getEnv(projectDir == null ? null : projectDir.toNioPath());
        if (oEnv.isEmpty()) {
            return;
        }
        val env = oEnv.get();
        Path exe;
        try {
            exe = Path.of(env.zigExecutable());
        } catch (InvalidPathException e) {
            Notifications.Bus.notify(new Notification("ZigBrains.Project",
                                                      "Invalid zig executable path: " + env.zigExecutable(),
                                                      NotificationType.ERROR));
            return;
        }
        if (!exe.isAbsolute()) {
            exe = toolchain.getLocation().resolve(exe);
        }
        if (!Files.exists(exe)) {
            Notifications.Bus.notify(new Notification("ZigBrains.Project",
                                                      "Zig executable path does not exist: " + env.zigExecutable(),
                                                      NotificationType.ERROR));
            return;
        }
        Path lib = null;
        override:
        if (state.overrideStdPath) {
            try {
                lib = Path.of(state.explicitPathToStd);
            } catch (InvalidPathException e) {
                Notifications.Bus.notify(new Notification("ZigBrains.Project",
                                                          "Invalid zig standard library path override: " + env.zigExecutable(),
                                                          NotificationType.ERROR));
                state.overrideStdPath = false;
                break override;
            }
            if (!lib.isAbsolute()) {
                lib = toolchain.getLocation().resolve(lib);
            }
            if (!Files.exists(lib)) {
                Notifications.Bus.notify(new Notification("ZigBrains.Project",
                                                          "Zig standard library path override does not exist: " + env.zigExecutable(),
                                                          NotificationType.ERROR));
                lib = null;
                state.overrideStdPath = false;
                break override;
            } else {
                state.explicitPathToStd = lib.toString();
            }
        }
        if (lib == null) {
            try {
                lib = Path.of(env.libDirectory());
            } catch (InvalidPathException e) {
                Notifications.Bus.notify(new Notification("ZigBrains.Project",
                                                          "Invalid zig standard library path: " + env.zigExecutable(),
                                                          NotificationType.ERROR));
                return;
            }
            if (!lib.isAbsolute()) {
                lib = toolchain.getLocation().resolve(lib);
            }
            if (!Files.exists(lib)) {
                Notifications.Bus.notify(new Notification("ZigBrains.Project",
                                                          "Zig standard library path does not exist: " +
                                                          env.zigExecutable(), NotificationType.ERROR));
                return;
            }
        }
        builder.zig_exe_path(exe.toString());
        builder.zig_lib_path(lib.toString());

    }
}

File: lsp/src/main/kotlin/com/falsepattern/zigbrains/project/toolchain/ToolchainZLSConfigProvider.kt
/*
 * This file is part of ZigBrains.
 *
 * Copyright (C) 2023-2025 FalsePattern
 * All Rights Reserved
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * ZigBrains is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, only version 3 of the License.
 *
 * ZigBrains is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with ZigBrains. If not, see <https://www.gnu.org/licenses/>.
 */

package com.falsepattern.zigbrains.project.toolchain

import com.falsepattern.zigbrains.lsp.config.SuspendingZLSConfigProvider
import com.falsepattern.zigbrains.lsp.config.ZLSConfig
import com.falsepattern.zigbrains.project.settings.zigProjectSettings
import com.intellij.notification.Notification
import com.intellij.notification.NotificationType
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.UserDataHolderBase
import com.intellij.openapi.util.io.toNioPathOrNull
import kotlin.io.path.pathString


class ToolchainZLSConfigProvider: SuspendingZLSConfigProvider {
    override suspend fun getEnvironment(project: Project, previous: ZLSConfig): ZLSConfig {
        val svc = project.zigProjectSettings
        var state = svc.state
        val toolchain = state.toolchain ?: project.suggestZigToolchain(UserDataHolderBase()) ?: return previous

        val env = toolchain.zig.getEnv(project).getOrElse { throwable ->
            throwable.printStackTrace()
            Notification(
                "zigbrains-lsp",
                "Failed to evaluate \"zig env\": ${throwable.message}",
                NotificationType.ERROR
            ).notify(project)
            return previous
        }

        val exe = env.zigExecutable.toNioPathOrNull() ?: run {
            Notification(
                "zigbrains-lsp",
                "Invalid zig executable path: ${env.zigExecutable}",
                NotificationType.ERROR
            ).notify(project)
            return previous
        }
        if (!exe.toFile().exists()) {
            Notification(
                "zigbrains-lsp",
                "Zig executable does not exist: $exe",
                NotificationType.ERROR
            ).notify(project)
            return previous
        }
        var lib = if (state.overrideStdPath && state.explicitPathToStd != null) {
            state.explicitPathToStd?.toNioPathOrNull() ?: run {
                Notification(
                    "zigbrains-lsp",
                    "Invalid zig standard library path override: ${state.explicitPathToStd}",
                    NotificationType.ERROR
                ).notify(project)
                null
            }
        } else null

        if (lib == null) {
            lib = env.libDirectory.toNioPathOrNull() ?: run {
                Notification(
                    "zigbrains-lsp",
                    "Invalid zig standard library path: ${env.libDirectory}",
                    NotificationType.ERROR
                ).notify(project)
                null
            }
        }
        if (lib == null)
            return previous

        return previous.copy(zig_exe_path = exe.pathString, zig_lib_path = lib.pathString)
    }
}